xdsl/transforms/csl_stencil_to_csl_wrapper.py

```diff
@@ -0,0 +1,232 @@
+            for result in apply_op.results:
+                if isa(result.type, stencil.TempType[TensorType[Attribute]]):
+                    z_dim_no_ghost_cells = max(
+                        z_dim_no_ghost_cells,
+                        result.type.get_element_type().get_shape()[0],
+                    )
+            for arg in op.args:
+                if isa(field_t := arg.type, stencil.FieldType[TensorType[Attribute]]):
+                    z_dim = max(z_dim, field_t.get_element_type().get_shape()[0])
+
+        # initialise module op
+        module_op = csl_wrapper.ModuleOp(
+            width=IntegerAttr(width, 16),
+            height=IntegerAttr(height, 16),
+            params={
+                "z_dim": IntegerAttr(z_dim, 16),
+                "pattern": IntegerAttr(neighbours + 1, 16),
+            },
+        )
+
+        self.initialise_layout_module(module_op)
+        module_op.update_program_block_args_from_layout()
+        module_op.set_program_name(op.sym_name)
+
+        # add func args to program_module block args
+        for block_arg in op.body.block.args:
+            module_op.program_module.block.insert_arg(
+                block_arg.type, len(module_op.program_module.block.args)
+            )
+
+        # replace func.return
+        func_return = op.body.block.last_op
+        assert isinstance(func_return, func.Return)
+        assert (
+            len(func_return.arguments) == 0
+        ), "Non-empty returns currently not supported"
+        rewriter.replace_op(func_return, csl.ReturnOp())
+
+        # set up main function and move func.func ops into this csl.func
+        main_func = csl.FuncOp(op.sym_name.data, ((), None))
+        rewriter.inline_block(
+            op.body.block,
+            InsertPoint.at_start(main_func.body.block),
+            module_op.exported_symbols,
+        )
+
+        # add main and empty yield to program_module
+        module_op.program_module.block.add_ops([main_func, csl_wrapper.YieldOp([], [])])
+
+        # replace (now empty) func by module wrapper
+        rewriter.replace_matched_op(module_op)
+
+    def get_csl_stencil_apply_ops(
+        self, op: func.FuncOp
+    ) -> Sequence[csl_stencil.ApplyOp]:
+        result: list[csl_stencil.ApplyOp] = []
+        for apply_op in op.body.walk():
+            if isinstance(apply_op, csl_stencil.ApplyOp):
+                result.append(apply_op)
+        return result
+
+    def initialise_layout_module(self, module_op: csl_wrapper.ModuleOp):
+        """Initialises the layout_module (wrapper block) by setting up (esp. stencil-related) program params"""
+
+        # extract layout module params as the function has linear complexity
+        param_width = module_op.get_layout_param("width")
+        param_height = module_op.get_layout_param("height")
+        param_x = module_op.get_layout_param("x")
+        param_y = module_op.get_layout_param("y")
+        param_pattern = module_op.get_layout_param("pattern")
+
```

Comment:
I think I'd personally prefer the style of:

```py
width, height, x, y, pattern = module_op.get_layout_params()
```

But alas, this is what we are doing now. Feel free to ignore my nitpicks

Reply:
That would prescribe an implementation-specific order and elevate pattern to always-required, hard-coding something I'm hoping not to hard-code.